source-code/core/app/src/createInlang.ts

```diff
@@ -0,0 +1,194 @@
+import type { InlangInstance } from "./api.js"
+import { ImportFunction, ResolveModulesFunction, resolveModules } from "@inlang/module"
+import { NodeishFilesystemSubset, InlangConfig, Message, tryCatch } from "@inlang/plugin"
+import { TypeCompiler } from "@sinclair/typebox/compiler"
+import { Value } from "@sinclair/typebox/value"
+import { InvalidConfigError } from "./errors.js"
+import { LintError, LintReport, lintMessages } from "@inlang/lint"
+import { createRoot, createSignal, createEffect } from "./solid.js"
+import { ReactiveMap } from "@solid-primitives/map"
+import { createReactiveQuery } from "./createReactiveQuery.js"
+
+const ConfigCompiler = TypeCompiler.Compile(InlangConfig)
+
+/**
+ * Creates an inlang instance.
+ *
+ * - Use `_import` to pass a custom import function for testing,
+ *   and supporting legacy resolvedModules such as CJS.
+ *
+ */
+export const createInlang = async (args: {
+	configPath: string
+	nodeishFs: NodeishFilesystemSubset
+	_import?: ImportFunction
+}): Promise<InlangInstance> =>
+	await createRoot(async () => {
+		const [initialized, markInitAsComplete] = createAwaitable()
+
+		// -- config ------------------------------------------------------------
+
+		const [config, setConfig] = createSignal<InlangConfig>()
+		createEffect(() => {
+			loadConfig({ configPath: args.configPath, nodeishFs: args.nodeishFs })
+				.then(setConfig)
+				.catch((err) => {
+					console.error("ERROR IN LOAD CONFIG ", err)
+				})
+		})
+		// TODO: create FS watcher and update config on change
+
+		console.log({ config })
+
+		// -- resolvedModules -----------------------------------------------------------
+
+		const [resolvedModules, setResolvedModules] =
+			createSignal<Awaited<ReturnType<ResolveModulesFunction>>>()
+
+		createEffect(() => {
+			const conf = config()
+			if (!conf) return
+
+			loadModules({ config: conf, nodeishFs: args.nodeishFs, _import: args._import })
+				.then((resolvedModules) => {
+					setResolvedModules(resolvedModules)
+					// TODO: handle `detectedLanguageTags`
+				})
+				.catch((err) => {
+					console.error("ERROR IN LOAD MODULES ", err)
+				})
+		})
+		// -- messages ----------------------------------------------------------
+
+		const [messages, setMessages] = createSignal<Message[]>()
+		createEffect(() => {
+			console.log("0 effect load messages is triggered")
+			const _resolvedModules = resolvedModules()
+			if (!_resolvedModules) return
+			//console.log(_resolvedModules.data.plugins.data)
+			makeTrulyAsync(
+				_resolvedModules.data.plugins.data.loadMessages({ languageTags: config()!.languageTags }),
```

Comment:
should we wrap all plugin defined functions to return a `Result`?

Reply:
How would an app fall back? Seems to me that an app should crash if loadMessages fails. But, also seems like we can change the behaviour in the future. We are not going to set @inlang/app to v1.0

Reply:
We would add the error to the plugin errors array to make it more consistent. This would be the same as an `ErrorDuringLintError`.
`loadMessages` could throw an error on the 3rd call of that function if the API endpoint is temporary down. Should the editor crash then? I think the app should decide that.

Reply:
> I think the app should decide that. 

Yeah, good point. 

@inlang/editor what do you think?

Reply:
I think that feels more consistent. So that would be a task for plugins.

Reply:
@NilsJacobsen and I agree that this is more consistent and is better handled within the app.